### PR TITLE
205 improve documentation on entry points

### LINKED ISSUE #205: Documentation of code out of sync in top level files
Code documentation of functions, function parameters and code comment has gone out of sync for fitting_benchmarking.py and fitbenchmark_one_problem.py and some of the functions it calls.

Fix:
* documentation for all function parameters
* check these are not out of date (and therefore currently wrong)
* add code explaining the code in do_benchmarking (in this function should problem_groups be problem_group and block be file?)

These two level python are particular important to be well documented since they form entry points to the other parts of FitBenchmarking.

Observed while testing #199

### PR DESCRIPTION
**Pull request by AndrewLister-STFC**
#### Description of Work

Fixes #205
Refactored do_benchmarking and edited docstrings to explain the code in a more understandable way.

Also contains commits for:
- Indentaion fixes
- Fix for broken mantid example
- Fix for expert example (I removed some pass through functions in a previous commit that were used by expert. I have put these back in and marked them as API functions.)

#### Testing Instructions

1. Check everything works as before
2. Read docstrings and see if it makes sense.

### CODE REVIEW

File: example_scripts/example_runScripts_expert.py [hunk 1]
```diff
@@ -112,34 +113,34 @@
...
-        # Generates accuracy and runtime normalised tables and summary tables
-        norm_acc_rankings, norm_runtimes, sum_cells_acc, sum_cells_runtime = generate_tables(group_results, minimizers)
+    # Creates the problem names with links to the visual display pages
+    # in rst
+    linked_problems = visual_pages.create_linked_probs(prob_results,
+                                                       group_name, results_dir)
 
-        # Creates an accuracy table
-        acc_tbl = create_acc_tbl(minimizers, linked_problems, norm_acc_rankings, use_errors, color_scale)
+    # Generates accuracy and runtime normalised tables and summary tables
```
Comment (Anders-Markvardsen) [inline example_scripts/example_runScripts_expert.py]:
```suggestion
    # Generates accuracy and runtime tables and summary tables
```

File: example_scripts/example_runScripts_mantid.py [hunk 2]
```diff
@@ -22,6 +22,7 @@
 fitbenchmarking_folder = os.path.abspath(os.path.join(current_path, os.pardir))
 scripts_folder = os.path.join(fitbenchmarking_folder, 'fitbenchmarking')
 sys.path.insert(0, scripts_folder)
+sys.path.insert(1, fitbenchmarking_folder)
```
Comment (Anders-Markvardsen) [inline example_scripts/example_runScripts_mantid.py]: question: is this needed if setup.py has been run?

**Comment by Anders-Markvardsen** [comment]
Thanks. As discussed removed for now not relevant block feature

**Comment by AndrewLister-STFC** [comment]
For the first issue I would expect it won't matter once the blocks are removed anyway (it'll change the code where the issue is raised.) but I'm not sure what could be causing the issue. It is working on my machine and Michaels when we've tried it.

> Thanks. As discussed removed for now not relevant block feature
I have just double checked the blocks on my laptop and it is also not working there so I'll remove it on monday.

**Comment by Anders-Markvardsen** [comment]
Running example_runScripts.py from anywhere remains broken, including for the location where a user most likely would like it from

```
In [11]: run example_runScripts.py

Running the benchmarking on the NIST_low_difficulty problem set

---------------------------------------------------------------------------
IOError                                   Traceback (most recent call last)
C:\Midlertiddig\fitbenchmarking\example_scripts\example_runScripts.py in <module>()
     96                            data_dir=data_dir,
     97                            use_errors=use_errors,
---> 98                            results_dir=results_dir)
     99
    100     print('\nProducing output for the {} problem set\n'.format(label))

C:\Midlertiddig\fitbenchmarking\fitbenchmarking\fitting_benchmarking.pyc in fitbenchmark_group(group_name, software_options, data_dir, use_errors, results_dir)
     36     logger.info("Loading minimizers from {0}".format(
     37         software_options['software']))
---> 38     minimizers, software = misc.get_minimizers(software_options)
     39
     40     # create list with blocks of paths to all problem definitions in data_dir

fitbenchmarking\utils\misc.py in get_minimizers(software_options)

fitbenchmarking\utils\options.pyc in get_option(options_file, option)

IOError: [Errno 2] No such file or directory: 'fitbenchmarking/fitbenchmarking_default_options.json'
```

This change degrades the user experience of FitBenchmarking.

**Comment by Anders-Markvardsen** [comment]
As a side when I run from the fitbenchmarking run directory: python setup.py install I get

```
copying sas\sascalc\pr\fit\expression.py -> build\lib\sas\sascalc\pr\fit
copying sas\sascalc\pr\fit\Loader.py -> build\lib\sas\sascalc\pr\fit
copying sas\sascalc\pr\fit\__init__.py -> build\lib\sas\sascalc\pr\fit
creating build\lib\benchmark_problems\NIST
creating build\lib\benchmark_problems\NIST\low_difficulty
copying benchmark_problems\NIST\low_difficulty\Chwirut1.dat -> build\lib\benchmark_problems\NIST\low_difficulty
copying benchmark_problems\NIST\low_difficulty\Chwirut2.dat -> build\lib\benchmark_problems\NIST\low_difficulty
copying benchmark_problems\NIST\low_difficulty\DanWood.dat -> build\lib\benchmark_problems\NIST\low_difficulty
copying benchmark_problems\NIST\low_difficulty\Gauss1.dat -> build\lib\benchmark_problems\NIST\low_difficulty
copying benchmark_problems\NIST\low_difficulty\Gauss2.dat -> build\lib\benchmark_problems\NIST\low_difficulty
copying benchmark_problems\NIST\low_difficulty\Lanczos3.dat -> build\lib\benchmark_problems\NIST\low_difficulty
copying benchmark_problems\NIST\low_difficulty\Misra1a.dat -> build\lib\benchmark_problems\NIST\low_difficulty
copying benchmark_problems\NIST\low_difficulty\Misra1b.dat -> build\lib\benchmark_problems\NIST\low_difficulty
creating build\lib\benchmark_problems\Neutron
error: can't copy 'benchmark_problems\Neutron\data_files': doesn't exist or not a regular file
```

which may not be related to this issue, but is a bug introduced into the install? Could you investigate where this bug originates from?

**Comment by tyronerees** [comment]
@Anders-Markvardsen -- @AndrewLister-STFC is looking at this in #240.  The fixes in #237 don't seem to work across all platforms, but as Travis doesn't run the example script at the moment this wasn't caught by the CI.

**Comment by AndrewLister-STFC** [comment]
> Running example_runScripts.py from anywhere remains broken, including for the location where a user most likely would like it from
> In [11]: run example_runScripts.py
> 
> Running the benchmarking on the NIST_low_difficulty problem set
> 
> ---------------------------------------------------------------------------
> IOError                                   Traceback (most recent call last)
> C:\Midlertiddig\fitbenchmarking\example_scripts\example_runScripts.py in <module>()
>      96                            data_dir=data_dir,
>      97                            use_errors=use_errors,
> ---> 98                            results_dir=results_dir)
>      99
>     100     print('\nProducing output for the {} problem set\n'.format(label))
> 
> C:\Midlertiddig\fitbenchmarking\fitbenchmarking\fitting_benchmarking.pyc in fitbenchmark_group(group_name, software_options, data_dir, use_errors, results_dir)
>      36     logger.info("Loading minimizers from {0}".format(
>      37         software_options['software']))
> ---> 38     minimizers, software = misc.get_minimizers(software_options)
>      39
>      40     # create list with blocks of paths to all problem definitions in data_dir
> 
> fitbenchmarking\utils\misc.py in get_minimizers(software_options)
> 
> fitbenchmarking\utils\options.pyc in get_option(options_file, option)
> 
> IOError: [Errno 2] No such file or directory: 'fitbenchmarking/fitbenchmarking_default_options.json'
> 
> This change degrades the user experience of FitBenchmarking.

The issue there is that you haven't specified an options file, and the default points to 'fitbenchmarking/fitbenchmarking_default_options.json', but I can change this to '../fitbenchmarking/fitbenchmarking_default_options.json' so it works from the example scripts directory instead. Originally it was using the location of the file and a relative path from there, which could be problematic also if things are installed in different places.

Perhaps the best way to solve this is to have the default as the current dir (e.g. './fitbenchmarking_default_options.json'), and move the options file there. Either way, this was a change introduced in  previous PR which was reviewed by @wathen, so a change to this should be a seperate issue.

To get past the fact that you have been unable to run it you should pass the options file in as an arg:
`python example_runScript.py ../fitbenchmarking_default_options.json`

File: example_scripts/example_runScripts_expert.py (hunk 1 above)
Comment (AndrewLister-STFC) [reply example_scripts/example_runScripts_expert.py]: Resolved in commit 10ca91a

File: example_scripts/example_runScripts_mantid.py (hunk 2 above)
Comment (AndrewLister-STFC) [reply example_scripts/example_runScripts_mantid.py]: This has been removed in another PR now

Comment (AndrewLister-STFC) [reply example_scripts/example_runScripts_mantid.py]: This has been removed in other PR

File: fitbenchmarking/fitting_benchmarking.py [hunk 3]
```diff
@@ -15,89 +15,70 @@
...
+
+    @returns :: array of fitting results for the problem group and
+                the path to the results directory
+    """
+
+    logger.info("Loading minimizers from {0}".format(
+        software_options['software']))
+    minimizers, software = misc.get_minimizers(software_options)
+
+    # create list with blocks of paths to all problem definitions in data_dir
```
Comment (Anders-Markvardsen) [inline fitbenchmarking/fitting_benchmarking.py]:
```suggestion
    # create list of paths to all problem definitions in data_dir
```

File: fitbenchmarking/utils/misc.py [hunk 4]
```diff
@@ -52,20 +52,18 @@ def get_minimizers(software_options):
         return minimizers, software
 
 
-def setup_fitting_problems(data_dir, group_name):
+def setup_fitting_problems(data_dir):
```
Comment (Anders-Markvardsen) [inline fitbenchmarking/utils/misc.py]: Could you nuke this method altogether in the code base and instead just call get_problem_files directly?

Comment (AndrewLister-STFC) [reply fitbenchmarking/utils/misc.py]: Yes, and done

**Pull request merged**